src/classes/UTIL_AbstractCommon_BATCH.cls

```diff
@@ -0,0 +1,453 @@
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Utilities
+* @description This class is meant to be extended by batch jobs requiring enhanced batch
+* functionality, including LDV Chunking Support, and support for automatically retrieving
+* subqueries for batches properly.  It is meant to switch between LDV Chunking mode or
+* non LDV Chunking mode, depending on the # of records in the object the batch is iterating
+* over.  The main limitation of this implementation is that it will not be possible to
+* provide an order to the records returned because it has to execute in a specific order
+* when running in LDV Chunking Mode.
+*
+* Due to the fact that constructors are not inherited by default, any constructors that
+* are desired need to be created using the signature as desired, and then in the constructor
+* body providing a super method and any params, e.g. super(batchSize);
+*
+* Where clause can be added by overriding getWhereClause() method
+*
+* Additional Where Clause can be configured by passing to constructor
+*
+* If Additional Where Clause is added, it can be validated by executing validateQuery().  If
+* it returns null, it is syntactically correct.  If it returns an Exception object, something about the
+* query is invalid.  The Additional Where and/or Where clause will be applied to the inner query for
+* each batch, if executing in LDV Chunking Mode, and the main batch query if executing in non LDV
+* Chunking mode.
+*
+* Future: Add exception handling, and logging support so implementing clases can inherit.
+* Recommend registering a default handler using an interface that can be switched out/overriden
+* as desired for specific use cases.
+*/
+public abstract with sharing class UTIL_AbstractCommon_BATCH implements Database.Batchable<SObject>, Database.Stateful {
+
+    private static final Integer MAX_LDV_CHUNK_SIZE = 99999;
```

Comment:
99K records doesn't seem like a lot for a chunk size if the job is iterating over 80M records. Is there a reason this couldn't be 1M or 10M? 

There's a balance to had over the total run time of the job versus the number of jobs that continue to chain over time. In a more reasonable org of 30M Opportunities, chunking at 99K record would mean that there 301 individual batch jobs executed, each with around 200 child batches (assuming batch size of 500). That's still a lot of actual batch jobs.

If this were to use a chunk size of 1M, then there would only be 30 batch jobs, each with around 2000 child batches. Depending on the complexity of the processing, one could expect each job to finish in around 1-5 hours.

The way that Salesforce dequeues jobs from the stack is generally to wait a little longer between each job as it's executed (when they chain together). That's going by some anecdotal discussions. I've personally seen where jobs that are in holding state will stay in a holding state longer as there are more jobs to process. Anyway, I think there's an advantage to a lower number of total jobs being executed with a balance to the number of records processed by each.

Reply:
So the max chunking size SF lists for their bulk api is 250K.  I refactored for that.  I am basing this on [PK Chunking Header](https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/async_api_headers_enable_pk_chunking.htm), but if you have counter/other info, I am just going by what is published.

Reply:
I do think if we want to go beyond that, I should refactor it into a constructor param, as it will become more of a tuning thing than a default/override situation.